pkg/secrethub/iterator.go

```diff
@@ -0,0 +1,57 @@
+package secrethub
+
+import (
+	"sync"
+
+	"github.com/secrethub/secrethub-go/internals/errio"
+)
+
+// Errors
+var (
+	IteratorDone = errio.Namespace("iterator").Code("done").Error("there are no more items left")
+)
+
+type paginator interface {
+	Next() ([]interface{}, error)
+}
+
+type iterator struct {
+	paginator    paginator
+	currentIndex int
+	items        []interface{}
+	mutex        *sync.Mutex
+}
+
+func newIterator(paginator paginator) iterator {
+	return iterator{
+		paginator:    paginator,
+		currentIndex: 0,
+		items:        nil,
+		mutex:        &sync.Mutex{},
+	}
+}
+
+func (it *iterator) next() (interface{}, error) {
+	it.mutex.Lock()
+	if it.items == nil || (len(it.items) > 0 && len(it.items) <= it.currentIndex) {
+		var err error
+		it.items, err = it.paginator.Next()
+		if err != nil {
+			it.mutex.Unlock()
```

Comment:
I'm not really seeing why a `defer it.mutex.Unlock()` wouldn't do the same trick here. Your commit message mentions something about a recursive call to `it.next`, but I do not see it.

Reply:
The recursive call is on line 45

Reply:
Ah dot it, totally missed that one. Could this benefit from a small wrapper function that does the mutex stuff and then calls this function so we don't have to deal with all those mutexes here?

Reply:
Created `nextUnsafe` so that the name makes clear that the function shouldn't be used directly. `next` obtains and releases the lock and calls `nextUnsafe`.